.github/workflows/macos.yml

```diff
@@ -47,26 +49,42 @@ jobs:
           path: ~/.ccache
           # We include the commit sha in the cache key, as new cache entries are
           # only created if there is no existing entry for the key yet.
-          key: ${{ runner.os }}-ccache-${{ github.sha }}
+          key: ${{ runner.os }}-${{ runner.arch }}-ccache-${{ github.sha }}
           # Restore any ccache cache entry, if none for
-          # ${{ runner.os }}-ccache-${{ github.sha }} exists.
+          # ${{ runner.os }}-${{ runner.arch }}-ccache-${{ github.sha }} exists.
           # Common prefix will be used so that ccache can be used across commits.
           restore-keys: |
-            ${{ runner.os }}-ccache
+            ${{ runner.os }}-${{ runner.arch }}-ccache
+
       - name: Set up Python version
         uses: actions/setup-python@v4
         with:
-          python-version: 3.8
+          python-version: ${{ matrix.os == 'macos-14' && '3.10' || '3.8' }}
+
       - name: Install dependencies
         run: |
           brew install ccache pkg-config
           # Install libomp 11.1.0 from old brew bottle for catalina (10.15).
           # Directly installing the Ruby formula will install for the current OS
           # https://github.com/microsoft/LightGBM/issues/4229
-          brew unlink libomp
-          curl -L -H "Authorization: Bearer QQ==" -o libomp-11.1.0.catalina.bottle.tar.gz \
-          https://ghcr.io/v2/homebrew/core/libomp/blobs/sha256:45a5aa653bd45bd5ff5858580b1a4670c4b5a51ea29d68d45a53f72f56010e05
-          brew install -f libomp-11.1.0.catalina.bottle.tar.gz
+          if [[ ${{ runner.arch}} == "X64"  ]]; then
+            brew unlink libomp
+            # x64 catalina (10.15) bottle
+            export LIBOMP_BOTTLE_HASH=45a5aa653bd45bd5ff5858580b1a4670c4b5a51ea29d68d45a53f72f56010e05
+          else # ARM64
+            # fix gfortran not found issue
+            ln -s $(which gfortran-13) /usr/local/bin/gfortran
+
+            # default Xcode 15.0.1 linker causes build issues, embree recommends clang <= 14 for arm64
+            sudo xcode-select -switch /Applications/Xcode_14.3.1.app
```

Comment:
No, it's not. I've moved it up. The default Xcode (15.0.1) for the macos-14 runner uses a new linker (ld-prime) which causes build issues. Newer Xcode versions fix these problems, but I decided to use Xcode 14.3.1 because the embree release mentions possible "EXEC_BAD_INSTRUCTION" runtime exceptions for Xcode >= 15: https://github.com/embree/embree/releases/tag/v4.3.1

I haven't seen these runtime exceptions being triggered. Added all of that as a code comment.